crates/packaging/src/https.rs

```diff
@@ -73,12 +90,19 @@ impl<'a> PackageMetadata<'a> {
             }
         };
 
-        // Next, check if there are misleading characters in the URL
-        if url
+        if let Some((index, ch)) = url
             .chars()
-            .any(|ch| MISLEADING_CHARACTERS_IN_URL.contains(&ch))
+            .enumerate()
+            .find(|(_, ch)| (!ch.is_alphanumeric() && !ALLOWED_URL_CHARACTERS.contains(ch)) || MISLEADING_CHARACTERS_IN_URL.contains(ch)) 
         {
-            return Err(UrlProblem::MisleadingCharacter);
+            // Check if there is an intentionally misleading url character
+            // Otherwise check if there is a possibly insecure character in the url
+            if MISLEADING_CHARACTERS_IN_URL.contains(&ch) {
+                return Err(UrlProblem::MisleadingCharacter);
+            } else {
+                return Err(UrlProblem::InsecureCharacter((ch.to_string(), index)));
+            }
+            
         }
```

Comment:
I think this'd be a bit easier to read/understand if rather than making this a single if, you split it out into two: one that's exactly like what was there before, and a second one after that to check this new condition.